src/source.cpp

```diff
@@ -53,71 +54,103 @@ namespace librealsense
 
     void frame_source::init(std::shared_ptr<metadata_parser_map> metadata_parsers)
     {
-        std::lock_guard<std::mutex> lock(_callback_mutex);
+        _supported_extensions = { RS2_EXTENSION_VIDEO_FRAME,
+                                  RS2_EXTENSION_COMPOSITE_FRAME,
+                                  RS2_EXTENSION_POINTS,
+                                  RS2_EXTENSION_DEPTH_FRAME,
+                                  RS2_EXTENSION_DISPARITY_FRAME,
+                                  RS2_EXTENSION_MOTION_FRAME,
+                                  RS2_EXTENSION_POSE_FRAME };
 
-        std::vector<rs2_extension> supported { RS2_EXTENSION_VIDEO_FRAME,
-                                               RS2_EXTENSION_COMPOSITE_FRAME,
-                                               RS2_EXTENSION_POINTS,
-                                               RS2_EXTENSION_DEPTH_FRAME,
-                                               RS2_EXTENSION_DISPARITY_FRAME,
-                                               RS2_EXTENSION_MOTION_FRAME,
-                                               RS2_EXTENSION_POSE_FRAME };
+        _metadata_parsers = metadata_parsers;
+    }
 
-        for (auto type : supported)
-        {
-            _archive[type] = make_archive(type, &_max_publish_list_size, metadata_parsers);
-        }
+    void frame_source::create_archive( archive_id id )
+    {
+        auto it = std::find( _supported_extensions.begin(), _supported_extensions.end(), id.second );
+        if( it == _supported_extensions.end() )
+            throw wrong_api_call_sequence_exception( "Requested frame type is not supported!" );
 
-        _metadata_parsers = metadata_parsers;
+        std::lock_guard< std::recursive_mutex > lock( _mutex );
+
+        _archive[id] = make_archive( id.second, &_max_publish_list_size, _metadata_parsers );
+        if( ! _archive[id] )
+            throw std::runtime_error( rsutils::string::from()
+                                      << "Failed to create archive of type " << get_string( id.second ) );
+
+        _archive[id]->set_sensor( _sensor );
     }
 
-    callback_invocation_holder frame_source::begin_callback()
+    callback_invocation_holder frame_source::begin_callback( archive_id id )
     {
-        return _archive[RS2_EXTENSION_VIDEO_FRAME]->begin_callback();
-//        return _archive[RS2_EXTENSION_DEPTH_FRAME]->begin_callback();
+        if( id.second >= RS2_EXTENSION_COUNT )
+            id.first = RS2_STREAM_COUNT; // For added extensions like GPU accelerated frames
```

Comment:
Needed?

Reply:
Yes needed.
When adding an extension we need to create the archive because `add_extension` is templated with the archive type. At that stage we don't know the stream type that will be used, but cannot keep the type for later allocation.
When using the allocated archive we need to `find` it with the same key.